packages/utils/src/partitionByKey.ts

```diff
@@ -39,90 +44,114 @@ export function partitionByKey<T, K, R>(
 export function partitionByKey<T, K>(
   stream: Observable<T>,
   keySelector: (value: T) => K,
-): [(key: K) => GroupedObservable<K, T>, Observable<K[]>]
+): [(key: K) => GroupedObservable<K, T>, Observable<KeyChanges<K>>]
 
 export function partitionByKey<T, K, R>(
   stream: Observable<T>,
   keySelector: (value: T) => K,
   streamSelector?: (grouped: Observable<T>, key: K) => Observable<R>,
-): [(key: K) => GroupedObservable<K, R>, Observable<K[]>] {
-  const groupedObservables$ = new Observable<Map<K, GroupedObservable<K, R>>>(
-    (subscriber) => {
-      const groups: Map<K, InnerGroup<T, K, R>> = new Map()
-
-      let emitted = false
-      let sourceCompleted = false
-      const sub = stream.subscribe(
-        (x) => {
-          const key = keySelector(x)
-          if (groups.has(key)) {
-            return groups.get(key)!.source.next(x)
-          }
-
-          const subject = new Subject<T>()
-
-          const res = shareLatest()(
-            (streamSelector || identity)(subject, key),
-          ) as GroupedObservable<K, R>
-          ;(res as any).key = key
-
-          const innerGroup: InnerGroup<T, K, R> = {
-            source: subject,
-            observable: res,
-            subscription: new Subscription(),
-          }
-          groups.set(key, innerGroup)
-
-          innerGroup.subscription = res.subscribe(
-            noop,
-            (e) => subscriber.error(e),
-            () => {
-              groups.delete(key)
-              subscriber.next(mapGroups(groups))
-
-              if (groups.size === 0 && sourceCompleted) {
-                subscriber.complete()
-              }
-            },
-          )
-
-          subject.next(x)
-          subscriber.next(mapGroups(groups))
```

Comment:
Note this was moved up before the `innerGroup.subscription` to cover the new test on synchronous completes. Otherwise it would emit first `remove` (on complete), and then `add` on this line.

It didn't break any other test, but I think we need to add more test around this area.